src/tasks.rs

```diff
@@ -185,3 +185,132 @@ impl TaskManager {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn create_and_update_task() {
+        let now = Instant::now();
+        let mut task = Task::new(Bytes::from_static(b"hello"));
+        assert!(task.create_at > now);
+        assert!(task.create_at < Instant::now());
+        assert!(matches!(task.code, TaskCode::UnknownError));
+        assert_eq!(task.data, Bytes::from_static(b"hello"));
+
+        task.update(TaskCode::Normal, &Bytes::from_static(b"world"));
+        assert!(matches!(task.code, TaskCode::Normal));
+        assert_eq!(task.data, Bytes::from_static(b"world"));
+    }
+
+    #[tokio::test]
+    async fn task_manager_add_new_task() {
+        let (tx, rx) = async_channel::bounded(1);
+        let task_manager = TaskManager::new(Duration::from_secs(1), tx);
+        let (id, _rx) = task_manager
+            .add_new_task(Bytes::from_static(b"hello"))
+            .await
+            .unwrap();
+        assert_eq!(id, 0);
+        {
+            let table = task_manager.table.read();
+            let task = table.get(&id).unwrap();
+            assert_eq!(task.data, Bytes::from_static(b"hello"));
+        }
+        let recv_id = rx.recv().await.unwrap();
+        assert_eq!(recv_id, id);
+
+        // add a new task
+        let (id, _rx) = task_manager
+            .add_new_task(Bytes::from_static(b"world"))
+            .await
+            .unwrap();
+        assert_eq!(id, 1);
+        {
+            let table = task_manager.table.read();
+            let task = table.get(&id).unwrap();
+            assert_eq!(task.data, Bytes::from_static(b"world"));
+        }
+        let recv_id = rx.recv().await.unwrap();
+        assert_eq!(recv_id, id);
+    }
+
+    #[tokio::test]
+    async fn task_manager_timeout() {
+        let (tx, _rx) = async_channel::bounded(1);
+        let task_manager = TaskManager::new(Duration::from_millis(1), tx);
+
+        // wait until this task timeout
+        let res = task_manager.submit_task(Bytes::from_static(b"hello")).await;
+        assert!(matches!(res.unwrap_err(), ServiceError::Timeout));
+    }
+
+    #[tokio::test]
+    async fn task_manager_too_many_request() {
+        let (tx, _rx) = async_channel::bounded(1);
+        // push one task into the channel to make the channel full
+        let _ = tx.send(0u32).await;
+        let task_manager = TaskManager::new(Duration::from_millis(1), tx);
+
+        // trigger too many request since the capacity is 0
```

Comment:
Why not `submit_task` twice to trigger this instead of manually operating on the `tx`?

Reply:
Because there is no consumer to process the task. So the first `submit_task` will result in `Timeout`.